src/Stripe.net/Infrastructure/JsonConverters/ExpandableFieldConverter.cs

```diff
@@ -0,0 +1,112 @@
+{
+    using System;
+    using System.Reflection;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// Converts a <see cref="ExpandableField{T}"/> to and from JSON.
+    /// </summary>
+    /// <typeparam name="T">Type of the field when expanded.</typeparam>
+    public class ExpandableFieldConverter<T> : StripeObjectConverter
+        where T : IHasId
+    {
+        /// <summary>
+        /// Gets a value indicating whether this <see cref="JsonConverter"/> can write JSON.
+        /// </summary>
+        /// <value>
+        ///     <c>true</c> if this <see cref="JsonConverter"/> can write JSON; otherwise, <c>false</c>.
+        /// </value>
+        public override bool CanWrite => true;
+
+        /// <summary>
+        /// Writes the JSON representation of the object.
+        /// </summary>
+        /// <param name="writer">The <see cref="JsonWriter"/> to write to.</param>
+        /// <param name="value">The value.</param>
+        /// <param name="serializer">The calling serializer.</param>
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            switch (value)
+            {
+                case null:
+                    serializer.Serialize(writer, null);
+                    break;
+
+                case IExpandableField expandableField:
+                    var valueToSerialize = expandableField.IsExpanded
+                        ? expandableField.ExpandedObject
+                        : expandableField.Id;
+                    serializer.Serialize(writer, valueToSerialize);
+                    break;
+
+                default:
+                    throw new JsonSerializationException(string.Format(
+                        "Unexpected value when converting ExpandableField. Expected IExpandableField, got {0}.",
+                        value.GetType().ToString()));
+            }
+        }
+
+        /// <summary>
+        /// Reads the JSON representation of the object.
+        /// </summary>
+        /// <param name="reader">The <see cref="JsonReader"/> to read from.</param>
+        /// <param name="objectType">Type of the object.</param>
+        /// <param name="existingValue">The existing value of object being read.</param>
+        /// <param name="serializer">The calling serializer.</param>
+        /// <returns>The object value.</returns>
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            var value = new ExpandableField<T>();
+
+            switch (reader.TokenType)
+            {
+                case JsonToken.Null:
+                    // Do nothing
+                    break;
+
+                case JsonToken.String:
+                    value.Id = (string)reader.Value;
+                    break;
+
+                case JsonToken.StartObject:
+                    var jObject = JObject.Load(reader);
```

Comment:
`JsonReader` is a forward-only stream reader. If the deserialization fails (which can happen with polymorphic types, if the API returns a type that the library isn't aware of), we want to take a second look at the JSON to make a last attempt at extracting the ID. This is why we load the JSON into a `JObject`, then create another reader from the `JObject` instance.